apps/src/netsim/NetSimRouterStatsTable.js

```diff
@@ -0,0 +1,287 @@
+
+  /**
+   * Last render time, in simulation-time.
+   * @type {number}
+   * @private
+   */
+  this.lastRenderTime_ = null;
+
+  /**
+   * Unix timestamp (local) of router creation time
+   * @type {number}
+   * @private
+   */
+  this.routerCreationTime_ = 0;
+
+  /**
+   * Total count of packets this router has received.
+   * @type {number}
+   * @private
+   */
+  this.totalPackets_ = 0;
+
+  /**
+   * Total count of packets this router has successfully processed.
+   * @type {number}
+   * @private
+   */
+  this.successfulPackets_ = 0;
+
+  /**
+   * Total size of all packets received by this router, in bits.
+   * @type {number}
+   * @private
+   */
+  this.totalData_ = 0;
+
+  /**
+   * Total size of all packets successfully processed by this router, in bits.
+   * @type {number}
+   * @private
+   */
+  this.successfulData_ = 0;
+
+  /**
+   * Maximum rate of data transfer (in bits per second)
+   * @type {number}
+   * @private
+   */
+  this.bandwidthLimit_ = 0;
+
+  /**
+   * Average rate of data transfer (in bits per second) over the last
+   * DATA_RATE_WINDOW_MS milliseconds.
+   * @type {number}
+   * @private
+   */
+  this.dataRate_ = 0;
+
+  /**
+   * Router's total memory capacity, in bits.
+   * @type {number}
+   * @private
+   */
+  this.totalMemory_ = 0;
+
+  /**
+   * Number of packets in the router's queue.
+   * @type {number}
+   * @private
+   */
+  this.queuedPackets_ = 0;
+
+  /**
+   * Current size of the router's packet queue, in bits.
+   * @type {number}
+   * @private
+   */
+  this.usedMemory_ = 0;
+
+  this.render({});
+};
+
+/**
+ * @param {RunLoop} runLoop
+ */
+NetSimRouterStatsTable.prototype.attachToRunLoop = function (runLoop) {
+  runLoop.render.register(this.render.bind(this));
+};
+
+/**
+ * Fill the root div with new elements reflecting the current state
+ * @param {RunLoop.Clock} clock
+ */
+NetSimRouterStatsTable.prototype.render = function (clock) {
+  if (!this.needsRender(clock)) {
+    return;
+  }
+
+  var renderedMarkup = $(markup({
+    uptime: this.getLocalizedUptime(),
+    queuedPackets: this.queuedPackets_,
+    totalPackets: this.totalPackets_,
+    successfulPackets: this.successfulPackets_,
+    totalData: this.totalData_,
+    successfulData: this.successfulData_,
+    bandwidthLimit: this.bandwidthLimit_,
+    dataRate: this.dataRate_,
+    totalMemory: this.totalMemory_,
+    usedMemory: this.usedMemory_
+  }));
+  this.rootDiv_.html(renderedMarkup);
+  this.lastRenderTime_ = clock.time;
+};
+
+/**
+ * @param {RunLoop.Clock} clock
+ * @returns {boolean} whether a render operation is needed.
+ */
+NetSimRouterStatsTable.prototype.needsRender = function (clock) {
+  return (!this.lastRenderTime_ ||
+      clock.time - this.lastRenderTime_ > MAX_RENDER_DELAY_MS);
+};
+
+/**
+ * Mark the router log data dirty, so that it will re-render on the
+ * next frame.
+ */
+NetSimRouterStatsTable.prototype.setNeedsRender = function () {
+  this.lastRenderTime_ = null;
```

Comment:
john carmack had a post about the dangers of overloading values with null to indicate an additional state, but i can't find it.  plus, everybody does it :)